torch/fx/passes/utils/matcher_utils.py

```diff
@@ -0,0 +1,230 @@
+from dataclasses import dataclass, field
+from collections import defaultdict
+import copy
+from torch.fx.graph import Graph
+from torch.fx.node import Node
+from torch.fx._compatibility import compatibility
+from typing import Dict, List, Set
+
+__all__ = ['SubgraphMatcher']
+
+
+@compatibility(is_backward_compatible=False)
+@dataclass
+class _InternalMatch():
+    # Node from which the match was found
+    anchor: Node
+    # Maps nodes in the pattern subgraph to nodes in the larger graph
+    nodes_map: Dict[Node, Node] = field(default_factory=dict)
+
+    # nodes in target graph that are matched placeholder in pattern
+    placeholder_nodes: List[Node] = field(default_factory=list)
+
+    # sink nodes in matched subgraph returned by output
+    returning_nodes: List[Node] = field(default_factory=list)
+
+    def __copy__(self):
+        return _InternalMatch(anchor=self.anchor, nodes_map=self.nodes_map.copy(),
+                              placeholder_nodes=self.placeholder_nodes.copy(),
+                              returning_nodes=self.returning_nodes.copy())
+
+@compatibility(is_backward_compatible=False)
+class SubgraphMatcher:
```

Comment:
Is this matcher only sound for functional graphs?

Reply:
No, it can also work with non-functional graph. 
The matcher is just implemented based on Node and edge (topological connectivity) of the Graph.